dask_match/tests/test_collection.py

```diff
@@ -139,12 +139,38 @@ def test_repr(ddf):
 
 
 def test_columns_traverse_filters(df, ddf):
-    result = optimize(ddf[ddf.x > 5].y)
+    result = optimize(ddf[ddf.x > 5].y, fuse=False)
     expected = ddf.y[ddf.x > 5]
 
     assert str(result) == str(expected)
 
 
+def test_optimize_fusion(ddf):
```

Comment:
Given the heavy logic in the fusion code my guess is that we'll want to think defensibly here a bit, and come up with some cases that might trip up a lesser algorithm.  I wouldn't be surprised, if in doing so, we uncover some failure cases of the code as it is written today.

Reply:
Absolutely agree. Once we are happy with the general organization of things, we will want very robust tests to help iterate on this algorithm.

Reply:
My hope is that the tests will be more robust than the code itself.  I would probably do this pretty early.  (There are some who would do this before writing any code really).  Hopefully these tests are just a bunch of situations that we think might get tricky.  They wouldn't depend on the actual implementation that we end up coming up with.

Reply:
> Turns out that it is pretty easy to break things outside of fusion while trying to "stress" fusion.

Can you say more here?  In what kinds of situations does this break?

Reply:
One of the things that "broke" was just the fact that `Series.name` was missing (trivial to fix).  However, I think the main challenge I'm running into is the fact that there are not many non-`Blockwise` operations to work with, and I'd like to be able to do something like `ddf1["a"] + ddf2["b"].sum()` to introduce a non-`Blockwise` `Expr` dependency to a `Blockwise` expression. Without being able to break up `Blockwise` chains like this, it is a bit challenging to write robust tests.

Reply:
Also, I believe if `optimize` was called within `compute` by default (even without fusion), some of the existing tests would be failing. I will probably try to find a way to test this hypothesis in a systematic way.

Reply:
Yes, that seems reasonable to me.

Regarding my earlier point about scalar-reduction results: It may be necessary for an `Expr` to "know" if it corresponds to a "scalar" result. Although we want most DataFrame/Series/Index-specific logic to live at the collection level, it seems important for the `Expr` to know when `divisions` are irrelevant. For example, if we had a `Expr.scalar` attribute that only returned `True` for expressions that we know will produce scalar values (using `Reduction._meta` for example), it would be relatively easy to broadcast the key for the scalar result to all partitions of a dependent `Blockwise` operation.

Reply:
> It may be necessary for an Expr to "know" if it corresponds to a "scalar" result

If we're looking for `is_scalar` then maybe we can infer this from `_meta`?

```python
def is_scalar(self):
    return is_scalar(self._meta)
```

Alternatively, maybe the thing we're looking for here isn't necessarily `is_scalar` but rather `is_broadcastable`.  For example in this case:

```python
df - df.mean()
```

The `df.mean()` value is likely a pandas.Series.  If that's what we're looking for then my suggestion above probably doesn't work.  In that case I'd agree that we probably want something like the following (which maybe you were suggesting above):

```python
class Expr:
    broadcastable = False

class Reduction(Expr):
    broadcastable = True
```

Are there other situations where we'll have broadcastable objects, or where Reductions aren't broadcastable?  I guess maybe there's the more general ACA case.  Maybe some of them will or won't be?

Reply:
I *think* you are right that I mostly care about whether or not the result is *broadcastable*. However, my understanding is that the only case that a result is "always" broadcastable is if it is also scalar, otherwise it sometimes depends on the specific operation (I may have this wrong).

>If we're looking for is_scalar then maybe we can infer this from _meta?

Yes, we can infer this from the meta, but my earlier profiling suggest that we should avoid materializing the meta when we don't **need** it. Therefore, I'd probably want to return `False` be default, and only add meta-inspection logic to the `Expr` classes that **might** be returning something that is broadcastable. Therefore, I'm pretty much proposing what yourr last code example, but with `Reduction.broadcastable` implemented as a property with simple logic to inspect `meta`.

>Are there other situations where we'll have broadcastable objects

Not sure if there are any other high-profile DataFrame operations that are broadcastable - good question.

>or where Reductions aren't broadcastable?

It depends on the collection you are broadcasting over, right?  I understand that you can broadcast a series over a DataFrame collection, but can you broadcast a series over a series collection? If we support split_out>1, there will certainly be general ACA results that can't be broadcasted.

Reply:
> but my earlier profiling suggest that we should avoid materializing the meta when we don't need it. 

I'm curious to learn more about this. We used to eagerly compute meta and it wasn't a problem.  My guess is that there is a way to make this work.  Probably this means tracking which expressions have expensive meta calculations (my guess is that this is only read_parquet) and then making sure that we're caching things appropriately.

 > Therefore, I'm pretty much proposing your last code example, but with Reduction.broadcastable implemented as a property with simple logic to inspect meta.

If the protocol generally is to have a property then I'm fine with that.  My guess is that Reduction is always true though.   
 If we're doing something like dask.Series + pandas.Series then that should fail some other check, other than broadcastable is my guess.  Maybe ACA is sometimes true and needs a proper property due to, for example, split_out.